lib/Dialect/ESI/ESITranslations.cpp

```diff
@@ -0,0 +1,225 @@
+//===- ESITranslations.cpp - ESI translations -------------------*- C++ -*-===//
+//
+// ESI translations:
+// - Cap'nProto schema generation
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/ESI/ESIDialect.h"
+#include "circt/Dialect/ESI/ESIOps.h"
+
+#include "circt/Dialect/RTL/Dialect.h"
+#include "circt/Dialect/SV/Dialect.h"
+
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/IR/BuiltinDialect.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/Translation.h"
+
+#include "llvm/ADT/DenseSet.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+#include <algorithm>
+
+using namespace mlir;
+using namespace circt::esi;
+
+//===----------------------------------------------------------------------===//
+// ESI Cosim Cap'nProto schema generation.
+//
+// Cosimulation in ESI is done over capnp. This translation walks the IR, finds
+// all the `esi.cosim` ops, and creates a schema for all the types.
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct ExportCosimSchema {
+  ExportCosimSchema(ModuleOp module, llvm::raw_ostream &os)
+      : module(module), os(os), diag(module.getContext()->getDiagEngine()),
+        unknown(UnknownLoc::get(module.getContext())) {
+    diag.registerHandler([this](Diagnostic &diag) -> LogicalResult {
+      if (diag.getSeverity() == DiagnosticSeverity::Error)
+        ++errorCount;
+      return failure();
+    });
+  }
+
+  /// Emit an ID in capnp format.
+  llvm::raw_ostream &emitId(llvm::hash_code id) {
+    return os << "@" << llvm::format_hex(id, /*width=*/16 + 2);
+  }
+
+  /// Emit the whole schema.
+  LogicalResult emit();
+
+  /// Collect the types for which we need to emit a schema. Output some metadata
+  /// comments.
+  LogicalResult visitEndpoint(CosimEndpoint);
+
+  /// Do we support emitting a schema for 'type'?
+  static bool isTypeSupported(Type type);
+
+  /// Emit a schema for a single int.
+  LogicalResult emitSchemaFor(IntegerType type);
+
+  /// Emit a struct name.
+  llvm::raw_ostream &emitName(Type type) { return os << "ESI_" << type; }
+
+private:
+  /// Intentation utils.
+  raw_ostream &indent() { return os.indent(currentIndent); }
+  void addIndent() { currentIndent += 2; }
+  void reduceIndent() { currentIndent -= 2; }
+  size_t currentIndent = 0;
+
+  ModuleOp module;
+  llvm::raw_ostream &os;
+  DiagnosticEngine &diag;
+  const Location unknown;
+  size_t errorCount = 0;
+
+  // All the `esi.cosim` input and output types encountered during the IR walk.
+  // This is NOT in a deterministic order!
+  llvm::DenseSet<Type> types;
+};
+} // anonymous namespace
+
+bool ExportCosimSchema::isTypeSupported(Type type) {
+  auto chanPort = type.dyn_cast<ChannelPort>();
+  if (chanPort) {
+    uint64_t innerHash = isTypeSupported(chanPort.getInner());
```

Comment:
Should this just `return isTypeSuppported(chanPort.getInner())`? I'm not sure why this boolean check is hashing the type in this case.

Reply:
Copy-paste error. I shouldn't submit PRs at the end of the night...